tripal/src/Plugin/TripalVocabulary/TripalDefaultVocabulary.php

```diff
@@ -0,0 +1,262 @@
+  /**
+   * An instance of the TripalLogger.
+   *
+   * @var \Drupal\tripal\Services\TripalLogger
+   */
+  protected $messageLogger = NULL;
+
+  /**
+   * A simple boolean to prevent Chado queries if the vocabulary isn't valid.
+   *
+   * @var bool
+   */
+  protected $is_valid = False;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    // Instantiate the TripalLogger
+    $this->messageLogger = \Drupal::service('tripal.logger');
+  }
+
+  /**
+   * Loads a Vocabulary record from Chado.
+   *
+   * This function queries the `cv` table of Chado to get the values
+   * for the vocabulary.
+   *
+   * @return
+   *   An associative array containing the columns of the `db1 table
+   *   of Chado or NULL if the db could not be found.
+   */
+  protected function loadVocab() {
+
+    $conn = \Drupal::service('database');
+
+    // Get the Chado `db` record for this ID space.
+    $query = $conn->select('tripal_terms_vocabs', 'vocabs')
+      ->condition('name', $this->getName(), '=')
+      ->fields('vocabs', ['name', 'label', 'url', 'idspaces']);
+    $result = $query->execute();
+    if ($result) {
+      return $result->fetchAssoc();
+    }
+    return NULL;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getIdSpaceNames() {
+    $cv = $this->loadVocab();
+    $idspaces = unserialize($cv['idspaces']);
+    return $idspaces;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getLabel() {
+    $cv = $this->loadVocab();
+    return $cv['label'];
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function recordExists() {
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function isValid() {
+
+    // Make sure the name of this collection does not exceeed the allowed size in Chado.
+    $name = $this->getName();
+    if (!empty($name) AND (strlen($name) > 255)) {
+      $this->messageLogger->error('TripalDefaultVocabular: The vocabulary name must not be longer than @size characters. ' +
+          'The value provided was: @value',
+          ['@size' => 255, '@value' => $this->getName()]);
+    }
+    $this->is_valid = TRUE;
+
+    return $this->is_valid;
+  }
+
+  /**
+   * Returns the namespace of the vocabulary
+   *
+   * This should be identical to the name of the collection, and
+   * therefore, there is no setter function.
+   *
+   * @return string $namespace
+   *   The namespace of the vocabulary.
+   */
+  public function getNameSpace() {
+    return $this->getName();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function destroy() {
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function setURL($url) {
+    // @todo there may be a problem in the future if we are able to
+    // associate borrowed terms with a vocabulary.  If we
+    // add the ID space of borrowed terms to a vocabulary then
+    // setting the URL will be incorrect for those ID spaces.
+
```

Comment:
I'm feeling really sad about the above comment. It seems there was something lost in translation between me + Josh. The original design for this API that me + Josh came up with does support borrowed terms. But when I went looking for it, this seems to have been changed at some point and I didn't catch it :-( You can see more about the original design here: https://tripaldoc.readthedocs.io/en/latest/design/vocab/file-structure.html and https://tripaldoc.readthedocs.io/en/latest/design/vocab.html.

This was intended to work by having multiple terms with the same ID Space and different vocabularies. 

For example, say you have the following term in it's primary vocabulary:

- Name: person
- ID Space: ABC
- Accession: 12345
- Vocabulary: The Core Ontology for People

This term was then hypothetically borrowed by the "Components of a Research Group Ontology". This would result in an additional term object in Tripal as so:

- Name: person
- ID Space: ABC
- Accession: 12345
- Vocabulary: Components of a Research Group Ontology

You would be able to tell that the term was borrowed because you could get the default vocabulary of the ID Space and see that it is different then the id space of this particular term.

Reply:
I copied this code from the ChadoVocabulary class and updated for the table I created in the public schema of Drupal.  So this problem is not unique to this class, but also the ChadoVocabulary class.  I think this should be added as a new issue and it should be addressed for both the TripalDefaultVocabulary and ChadoVocabulary classes.

Reply:
Can we fix this later so we can get this merged in so I can get the testing done for PR #1413?

Reply:
Ok, I've created an issue here: https://github.com/tripal/tripal/issues/1421 for this so we don't forget.